cEP-0018.md

```diff
@@ -0,0 +1,216 @@
+   traversing the AST using methods such as `get_next_node` and
+   `get_previous_node` (inside `coantlib`).
+7. The bears dependant on the `coantlib` will reside in the same repository as
+   the coantlib. All of them will be in a different package, the coant-bears
+   package, all of them installable together via a single option in the
+   setup.py ([similar to this](https://stackoverflow.com/q/42859852)).
+8. The bears will have an option to inherit another language specific class to
+   utilise the high-level api (for e.g in place of ASTBear, they inherit
+   Py3ASTBear).
+
+## Management of the new `coala-ast` repository
+
+Managing a new repository is a heavy task, and this will be highly automated as
+follows:
+
+1. The parsers in coala-antlr/coantparsers would be generated and
+   pushed via ci-builds whenever a new commit is pushed.
+2. The cib tool can be enhanced to deal with the installation of bears that
+   require only some specified parsers (for e.g a PyUnusedVarBear would
+   only require parser for python).
+3. The cib tool can also trigger specialised builds and download the newly
+   generated parser on the fly.
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations within coantlib:
+
+```python
+import antlr4
+import coantparsers
+
+from coalib.bearlib.languages.definitions import *
+from coalib.bearlib.languages.Language import parse_lang_str, Languages
+
+class ASTLoader():
+    mapping = {
+        C : [coantparsers.clexer, coantparsers.cparser],
+        Python : [coantparsers.pylexer, coantparsers.pyparser],
+        JavaScript : [coantparsers.jslexer, coantparsers.jsparser],
+        ...
+    }
+
+    @staticmethod
+    def loadFile(file,filename,lang = 'auto'):
+        """
+        Loads file's AST into memory
+        """
+        if(lang == 'auto'):
+          ext = get_extension(filename)
+        else:
+          ext = Languages(parse_lang_str(lang)[0])[0]
+        if ext is None:
+            raise FileExtensionNotSupported('Unknown File Type')
+        inputStream = file.readlines()
+        lexer = mapping[ext][0](inputStream)
+        parser = mapping[ext][1](lexer)
+        return parser
+
+class ASTWalker():
+    treeRoot = None
+    tree = None
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """
+        Modify the internal tree variable and return the next node
+        """
+
+    def get_prev_node(self):
+        """
+        Modify the internal tree variable and return prev node
+        """
+
+    def reset(self):
+        """
+        Reset the internal tree variable to tree root
+        """
+
+    def is_terminal_node(node):
+        """
+        Determine if the node is terminal
+        """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+    ...
+```
+
+```python
+class Py3ASTBear(ASTBear):
```

Comment:
Nit issue, Not serious for this time.
This design can be updated this way since most of the syntaxes are duplicated.
What we need to do is overriding or add an additional method for Python3 or Python4.

```
class PythonBaseASTBear(ASTBear):
       pass

class Python2ASTBear(ASTBear):
       pass

class Python3ASTBear(ASTBear):
       pass
```